eox_core/edxapp_wrapper/backends/enrollment_h_v1.py

```diff
@@ -28,24 +26,40 @@ def create_enrollment(*args, **kwargs):
     course_id = kwargs.get('course_id')
     mode = kwargs.get('mode')
     is_active = kwargs.get('is_active', True)
+    force_registration = kwargs.get('force_registration', False)
     enrollment_attributes = kwargs.get('enrollment_attributes', None)
     validation_errors = check_edxapp_enrollment_is_valid(*args, **kwargs)
     if validation_errors:
         return None, [", ".join(validation_errors)]
     if email:
-        match = get_user_model().objects.filter(email=email).first()
+        match = User.objects.filter(email=email).first()
         if match is None:
             raise APIException('No user found with that email')
         else:
             username = match.username
 
     try:
         enrollment = api._data_api().create_course_enrollment(username, course_id, mode, is_active)
-        if enrollment_attributes is not None:
-            api.set_enrollment_attributes(username, course_id, enrollment_attributes)
-    except (UserNotFoundError, InvalidEnrollmentAttribute, CourseNotFoundError, CourseEnrollmentFullError, CourseEnrollmentClosedError, CourseEnrollmentExistsError) as e:
-        enrollment = None
-        raise APIException(repr(e))
+    except CourseEnrollmentExistsError as e:
+        if force_registration:
+            enrollment = api._data_api().update_course_enrollment(username, course_id, mode, is_active)
+        else:
+            raise APIException(repr(e) + ", use force_registration to update the existing enrollment")
+    except Exception as e:
+        if force_registration:
+            try:
+                course_key = CourseKey.from_string(course_id)
+                user = User.objects.get(username=username)
+                enrollment = CourseEnrollment.enroll(user, course_key, check_access=False)
+                api._data_api()._update_enrollment(enrollment, is_active=is_active, mode=mode)
+            except Exception as e:
+                raise APIException(repr(e))
+        else:
+            raise APIException(repr(e))
+
+    if enrollment_attributes is not None:
+        api.set_enrollment_attributes(username, course_id, enrollment_attributes)
```

Comment:
Any chance that this is the line that fails?

Reply:
not really